File: src/main/java/com/codingforcookies/betterrecords/item/ItemRecord.java
package com.codingforcookies.betterrecords.item;

import com.codingforcookies.betterrecords.api.record.IRecord;
import com.codingforcookies.betterrecords.api.wire.IRecordWireHome;
import com.codingforcookies.betterrecords.common.packets.PacketHandler;
import com.codingforcookies.betterrecords.common.util.BetterUtils;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.util.text.translation.I18n;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

import java.util.List;

public class ItemRecord extends BetterItem implements IRecord {

    public ItemRecord(String name) {
        super(name);
        setMaxStackSize(1);
    }

    @Override
    @SideOnly(Side.CLIENT)
    public String getItemStackDisplayName(ItemStack itemStack) {
        if(itemStack.getTagCompound() != null && itemStack.getTagCompound().hasKey("local")) return itemStack.getTagCompound().getString("local");
        else return I18n.translateToLocal(getUnlocalizedName() + ".name");
    }

    @Override
    @SideOnly(Side.CLIENT)
    public void addInformation(ItemStack itemStack, EntityPlayer player, List<String> tooltip, boolean advanced) {
        if(itemStack.getTagCompound() != null && itemStack.getTagCompound().hasKey("author")) tooltip.add(BetterUtils.getTranslatedString("item.record.by") + ": " + itemStack.getTagCompound().getString("author"));
        if(itemStack.getTagCompound() != null && itemStack.getTagCompound().hasKey("size")) tooltip.add(BetterUtils.getTranslatedString("item.record.size") + ": " + itemStack.getTagCompound().getString("size") + "mb");
        if(itemStack.getTagCompound() != null && itemStack.getTagCompound().hasKey("repeat") && itemStack.getTagCompound().getBoolean("repeat")){
            tooltip.add("");
            tooltip.add("\247e" + BetterUtils.getTranslatedString("item.record.repeatenabled"));
        }
    }

    @Override
    public boolean isRecordValid(ItemStack itemStack) {
        return itemStack.getTagCompound() != null && itemStack.getTagCompound().hasKey("name");
    }

    @Override
    public void onRecordInserted(IRecordWireHome wireHome, ItemStack itemStack) {
        PacketHandler.sendRecordPlayToAllFromServer(wireHome.getTileEntity().getPos().getX(), wireHome.getTileEntity().getPos().getY(), wireHome.getTileEntity().getPos().getZ(), wireHome.getTileEntity().getWorld().provider.getDimension(), wireHome.getSongRadius(), itemStack.getTagCompound().getString("name"), itemStack.getTagCompound().getString("url"), itemStack.getTagCompound().getString("local"), itemStack.getTagCompound().hasKey("repeat") ? itemStack.getTagCompound().getBoolean("repeat") : false, itemStack.getTagCompound().hasKey("shuffle") ? itemStack.getTagCompound().getBoolean("shuffle") : false);
    }
}

File: src/main/kotlin/com/codingforcookies/betterrecords/item/ItemRecord.kt
package com.codingforcookies.betterrecords.item

import com.codingforcookies.betterrecords.api.record.IRecord
import com.codingforcookies.betterrecords.api.wire.IRecordWireHome
import com.codingforcookies.betterrecords.client.old.sound.Sound
import com.codingforcookies.betterrecords.network.PacketHandler
import com.codingforcookies.betterrecords.network.PacketRecordPlay
import net.minecraft.client.resources.I18n
import net.minecraft.client.util.ITooltipFlag
import net.minecraft.item.ItemStack
import net.minecraft.world.World
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly
import net.minecraft.util.text.translation.I18n as ServerI18n

open class ItemRecord(name: String) : ModItem(name), IRecord {

    init {
        maxStackSize = 1
    }

    override fun isRecordValid(itemStack: ItemStack) =
        itemStack.hasTagCompound() && itemStack.tagCompound!!.hasKey("name")

    override fun onRecordInserted(wireHome: IRecordWireHome, itemStack: ItemStack) {
        itemStack.tagCompound?.let {
            PacketHandler.sendToAll(PacketRecordPlay(
                    wireHome.tileEntity.pos,
                    wireHome.tileEntity.world.provider.dimension,
                    wireHome.songRadius,
                    it.getBoolean("repeat"),
                    it.getBoolean("shuffle"),
                    sound = Sound().setInfo(
                            it.getString("name"),
                            it.getString("url"),
                            it.getString("local")
                    )
            ))
        }
    }

    @SideOnly(Side.CLIENT)
    override fun addInformation(stack: ItemStack, worldIn: World?, tooltip: MutableList<String>, flagIn: ITooltipFlag) {
        stack.tagCompound?.let {
            tooltip += I18n.format("item.betterercords:record.desc.by", it.getString("author"))
            tooltip += I18n.format("item.betterercords:record.desc.size", it.getInteger("size"))
            if (it.getBoolean("repeat")) {
                tooltip += ""
                tooltip += "\u00a7e" + I18n.format("item.betterercords:record.desc.repeat")
            }
        }
    }

    override fun getItemStackDisplayName(stack: ItemStack): String =
            if (stack.hasTagCompound() && stack.tagCompound!!.hasKey("local")) {
                stack.tagCompound!!.getString("local")
            } else {
                ServerI18n.translateToLocal("$unlocalizedName.name")
            }
}
